lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs

```diff
@@ -0,0 +1,131 @@
+    /// kept as <see cref="IEnumerable{T}" />.
+    /// </summary>
+    public abstract class AbstractMetric : IMetricsInfo
+    {
+        private readonly IMetricsInfo _info;
+
+        private readonly long _value;
+
+        /// <summary>
+        /// Protected constructor of abstract metric.
+        /// </summary>
+        /// <param name="info">Meta-data for the metric</param>
+        /// <param name="value">The value of the metric. It is assumed that user 
+        /// or application has converted float/double values to long via some logic.</param>
+        protected AbstractMetric(IMetricsInfo info, long value)
+        {
+            _info = info;
+            _value = value;
+        }
+
+        /// <summary>
+        /// Name of the metric.
+        /// </summary>
+        public string Name
+        {
+            get { return _info.Name; }
+        }
+
+        /// <summary>
+        /// Description of the metric.
+        /// </summary>
+        public string Description
+        {
+            get { return _info.Description; }
+        }
+
+        /// <summary>
+        /// Meta-data of the metric.
+        /// </summary>
+        protected IMetricsInfo Info
+        {
+            get { return _info; }
+        }
+
+        /// <summary>
+        /// Value of the metric. Always stored as long. Immutable metrics of 
+        /// type integers, float, double, bool etc. are all converted to long before storing
+        /// them as immutable metrics. For double and float we can multiply by 10^n before 
+        /// converting to immutable Abstract metric. Exact logic of this conversion is left 
+        /// to user/application implementing <see cref="IMetricsSource"/>.
+        /// </summary>
+        public long Value
+        {
+            get { return _value; }
+        }
+
+        /// <summary>
+        /// Type of metric - counter or gauge. Filled in by exact 
+        /// metric type.
+        /// </summary>
+        public abstract MetricType TypeOfMetric { get; }
+
+        /// <summary>
+        /// Accepts a visitor interface
+        /// </summary>
+        /// <param name="visitor">Metrics visitor interface.</param>
+        public abstract void Visit(IMetricsVisitor visitor);
+
+        /// <summary>
+        /// String representation of a metric for display.
+        /// </summary>
+        /// <returns>The string representation of the metric.</returns>
+        public override string ToString()
+        {
+            return "Metric Information: " + _info.ToString() + ", Metric Value: " + Value;
+        }
+
+        /// <summary>
+        /// Checks whether two metrics are equal. Relies on Equals 
+        /// function of <see cref="IMetricsInfo"/> implementations.
+        /// </summary>
+        /// <param name="obj">Object to compare against.</param>
+        /// <returns>True if both represent the same metric.</returns>
+        public override bool Equals(object obj)
```

Comment:
In general name and description are supposed to identify the metric uniquely. I am wondering if we want inter-operability between int-long and float-double in which case its ok to not include type.